query/src/main/java/org/infinispan/query/clustered/commandworkers/CQCreateEagerQuery.java

```diff
@@ -18,38 +18,35 @@
 final class CQCreateEagerQuery extends CQWorker {
 
    @Override
-   QueryResponse perform(BitSet segments) {
+   CompletionStage<QueryResponse> perform(BitSet segments) {
       SearchQueryBuilder query = queryDefinition.getSearchQuery();
       setFilter(segments);
 
-      NodeTopDocs nodeTopDocs = (query.hasEntityProjection()) ? collectKeys(query) : collectProjections(query);
-      if (nodeTopDocs == null) {
-         return new QueryResponse(0);
-      }
-      return new QueryResponse(nodeTopDocs);
+      CompletionStage<NodeTopDocs> nodeTopDocs = (query.hasEntityProjection()) ? collectKeys(query) : collectProjections(query);
+
+      return nodeTopDocs.thenApply(QueryResponse::new);
    }
 
-   private NodeTopDocs collectKeys(SearchQueryBuilder query) {
-      LuceneSearchResult<DocumentReference> queryResult = query.documentReference().fetchAll();
-      if (queryResult.totalHitCount() == 0L) {
-         return null;
-      }
-
-      Object[] keys = queryResult.hits().stream()
-            .map(hit -> hit.id())
-            .map(id -> stringToKey(id))
-            .toArray(Object[]::new);
-      return new NodeTopDocs(cache.getRpcManager().getAddress(), queryResult.topDocs(), keys, null);
+   private CompletionStage<NodeTopDocs> collectKeys(SearchQueryBuilder query) {
+      return blockingManager.supplyBlocking(() -> query.documentReference().fetchAll(), this)
```

Comment:
I would recommend putting something a bit more descriptive for the second argument, maybe `CQCreateEagerQuery#collectKeys` ?